Mathlib/Data/Option/Defs.lean

```diff
@@ -57,3 +59,71 @@ def melim {α β : Type _} {m : Type _ → Type _} [Monad m] (x : m (Option α))
 /-- A monadic analogue of `Option.getD`. -/
 def mgetD {α : Type _} {m : Type _ → Type _} [Monad m] (x : m (Option α)) (y : m α) : m α :=
   melim x y pure
+
+variable {α : Type _} {β : Type _}
+
+-- Porting note: Would need to add the attribute directly in `Init.Prelude`.
+-- attribute [inline] Option.isSome Option.isNone
+
+-- Porting note: `Option.elim` already present with different argument order.
+/-- An elimination principle for `option`. It is a nondependent version of `option.rec`. -/
+@[simp]
+protected def elim' (b : β) (f : α → β) : Option α → β
+  | some a => f a
+  | none => b
+
+theorem is_none_iff_eq_none {o : Option α} : o.isNone = true ↔ o = none :=
+  ⟨Option.eq_none_of_isNone, fun e => e.symm ▸ rfl⟩
+
+theorem mem_some_iff {α : Type _} {a b : α} : a ∈ some b ↔ b = a := by simp
+
+/-- `o = none` is decidable even if the wrapped type does not have decidable equality.
+This is not an instance because it is not definitionally equal to `option.decidable_eq`.
+Try to use `o.isNone` or `o.isSome` instead.
+-/
+@[inline]
+def decidableEqNone {o : Option α} : Decidable (o = none) :=
+  decidable_of_decidable_of_iff is_none_iff_eq_none
+
+instance decidableForallMem {p : α → Prop} [DecidablePred p] : ∀ o : Option α, Decidable (∀ a ∈ o, p a)
+  | none => isTrue (by simp [false_imp_iff])
+  | some a => if h : p a then isTrue fun o e => some_inj.1 e ▸ h else isFalse <| mt (fun H => H _ rfl) h
+
+instance decidableExistsMem {p : α → Prop} [DecidablePred p] : ∀ o : Option α, Decidable (∃ a ∈ o, p a)
+  | none => isFalse fun ⟨a, ⟨h, _⟩⟩ => by cases h
+  | some a => if h : p a then isTrue <| ⟨_, rfl, h⟩ else isFalse fun ⟨_, ⟨rfl, hn⟩⟩ => h hn
+
+/-- Inhabited `get` function. Returns `a` if the input is `some a`, otherwise returns `default`. -/
+@[reducible]
+def iget [Inhabited α] : Option α → α
+  | some x => x
+  | none => default
+
+-- Porting note: Illegal simp attribute, left side is `a`
+-- @[simp]
+theorem iget_some [Inhabited α] {a : α} : (some a).iget = a :=
+  rfl
+
+@[simp]
+theorem mem_to_list {a : α} {o : Option α} : a ∈ toList o ↔ a ∈ o := by
+  cases o <;> simp [toList, eq_comm]
+
+-- lift f
+instance lift_or_get_comm (f : α → α → α) [h : IsCommutative α f] : IsCommutative (Option α) (liftOrGet f) :=
+  ⟨fun a b => by cases a <;> cases b <;> simp [lift_or_get, IsCommutative.comm]⟩
+
+instance lift_or_get_assoc (f : α → α → α) [h : IsAssociative α f] : IsAssociative (Option α) (liftOrGet f) :=
+  ⟨fun a b c => by cases a <;> cases b <;> cases c <;> simp [lift_or_get, IsAssociative.assoc]⟩
```

Comment:
Why does `simp` fail if I instead supply `h.assoc`?

Reply:
At this point, you can remove the `[h : ...]` `h` identifer. I also ran into this, you like saw my zulip thread.